src/lib/hooks/use-safe-client.ts

```
@@ -0,0 +1,115 @@
+
+      if (!address) {
+        setSafes([])
+        return
+      }
+
+      const { safes } = await apiKit.getSafesByOwner(address)
+      if (safes.length === 1) setSafeAddress(safes[0])
+
+      setSafes(safes)
+    }
+    fetchSafes()
+  }, [apiKit, address])
+
+  const validSafeSignature = async (typedData: EIP712TypedData) => {
+    if (!connectorClient || !address) return
+
+    const protocolKit = await Safe.init({
+      provider: connectorClient.transport,
+      safeAddress: address,
+    })
+
+    try {
+      const messageHash = hashSafeMessage(typedData)
+      const safeMessageHash = await protocolKit.getSafeMessageHash(messageHash)
+      const safeMessage = await apiKit.getMessage(safeMessageHash)
+      const isValid = await protocolKit.isValidSignature(
+        messageHash,
+        safeMessage.preparedSignature,
+      )
+      return isValid ? safeMessage.preparedSignature : null
+    } catch {
+      // apiKit.getMessage throws if the safeMessageHash is not found
+      // e.g. the message does not exist and the signature is not valid
+      return null
+    }
+  }
+
+  const signTypedData = async (typedData: EIP712TypedData) => {
+    if (!protocolKit || !safeAddress || !address) return
+
+    const safeMessage = await protocolKit.signMessage(
+      protocolKit.createMessage(typedData),
+      SigningMethod.ETH_SIGN_TYPED_DATA_V4,
+    )
+
+    const messageHash = hashSafeMessage(typedData)
+    const safeMessageHash = await protocolKit.getSafeMessageHash(messageHash)
+    const signature = safeMessage.getSignature(address) as EthSafeSignature
+    try {
+      const confirmedMessage = await apiKit.getMessage(safeMessageHash)
+      if (confirmedMessage) {
+        // Message exists and is already signed
+        await apiKit.addMessageSignature(
+          safeMessageHash,
+          buildSignatureBytes([signature]),
+        )
+      }
+    } catch (e) {
+      console.error('e', e)
+      // Message not created yet
```

Is this the only error that could happen or could there be a specific error code for this?

---

I didn't see any specific error codes returned (just JS errors) and it's unfortunate that it breaks hard/throws instead of returning `null` or similar. I believe the `addMessageSignature` could throw as well if the message already has its signature, but it won't end up creating an additional message in the `catch` as the message is based on the `hash`/won't duplicate.